### ISSUE TITLE
Logging HTTP requests

### ISSUE THREAD

**Issue by AlexanderKaraberov** [Moya/Moya#220]
I don't know if this is relevant, but Moya adds another level of abstraction on top of Alamofire which is on top of NSURLSession, so with all this debug logging of all the network stuff can be hard. I think may be we can implement something similar to incredibly useful https://github.com/AFNetworking/AFNetworkActivityLogger in such a way it fits Moya's architecture? IMO it will be very useful. E.g. add enableLogging flag to Endpoint with print() calls or something better?

**Comment by ashfurrow** [comment]
Sounds like a good idea – doesn't sound too difficult, either. We could hook into the existing provider initialization, if that makes sense, or use NSNotifications. Thoughts?

**Comment by AlexanderKaraberov** [comment]
If we add something like `print("\(applicationName): \(date) Request Headers:\(request.request!.allHTTPHeaderFields)")` inside `func request(token: T, completion: MoyaCompletion)` it will be too intrusive. Or you mean smth different in "hook into the existing provider initialization".  So `NSNotification` sounds like a good idea.

**Comment by ashfurrow** [comment]
Aye, so this is what I was thinking:

A user could pass in an endpoint resolver or other callback we already have – the one they pass in would log stuff. So that's how it could work now. With a new library, we could call a function that curries back a function to pass in. 

```swift
Provider<TargetEnum>(endpointResolver: MoyaNetworkLogger.StandardLogger)
... or ...
Provider<TargetEnum>(endpointResolver: MoyaNetworkLogger.StandardLogger { target -> NSURLRequqest in
    return /* our custom request, which the logger will then log before passing onto Moya */
})
```

Make sense?

**Comment by AlexanderKaraberov** [comment]
Great idea. We can go with this solution. What about logging a response?

**Comment by ashfurrow** [comment]
Hmm, good question. There's no existing way to log these at all – we may have to add yet another callback to the provider :cry:

**Comment by ashfurrow** [comment]
What if we added a property to the provider – could be constant or not – that conformed to a `Logger` protocol or something, and it gets callbacks about requests and responses. We could provide a basic logger, and people would be free to implement the protocol and provide their own. Thoughts?

**Comment by AlexanderKaraberov** [comment]
Just to be sure I understood you correctly. We will have a `protocol Logger` with some methods. Also we would have some kind of trait/typeclass (`protocol extension`) with default implementations of the logging methods. And we will use it as some kind of delegate?

**Comment by ashfurrow** [comment]
Hey, this is _almost_ correct. 

```swift
protocol Logger {
    // logging callbacks
}
```

Then provide a default:

```swift
struct PlainLogger: Logger {
    // Do basic logging, basically print() all the requests/responses
}
```

Then have a new variable on the `MoyaProvider`.

```swift
var logger: Logger?
```

Then in our `request` method in the `MoyaProvider` itself, we can invoke the callbacks:

```swift
logger?.logRequest(request)
...
logger?.logResponse(response)
```

I'm thinking the logger doesn't touch the `MoyaTarget` at all – just the HTTP requests. Does that makes sense?

**Comment by AlexanderKaraberov** [comment]
@ashfurrow  Why don't you want to make it an `Endpoint` property as `let logger: Logger`, init it in the initializer with default `PlainLogger` and then add convenience method for changing it (if user wants some custom logging for particular requests) like `endpointByAddingLogger...`. In this case we will have a flexible logging behaviour which will fit the entire Moya archcitecture.

**Comment by AlexanderKaraberov** [comment]
@ashfurrow  I understood my mistake, so I deleted my response:)
My only question - why do you want to make it `var` not `let`. I don't really want mutable properties inside `Provider`

**Comment by ashfurrow** [comment]
Hmm, yeah. I suppose it makes more sense to be `let` :smile_cat: 

I think it should be optional, though, and default to `nil`. Many (most?) users won't want HTTP logging.

**Comment by AlexanderKaraberov** [comment]
@ashfurrow "Many (most?) users won't want HTTP logging" - reasonable :blush: 
Ok, so we will have a nullable property and no logging by default, a `PlainLogger` (better `MoyaLogger` or smth. like this), which we have to include in the documentation and `Logger` protocol, so users can provide their own logging implementations if they want.

**Comment by ashfurrow** [comment]
Yeah!

**Comment by AlexanderKaraberov** [comment]
Great.  hope to implement this soon, when I will have more free time.

**Comment by ashfurrow** [comment]
Awesome, looking forward to your pull request! Don't worry about getting it in immediately – we've all got lives outside or GitHub 😸

**Comment by aschuch** [comment]
This can be closed with the release of 3.0.
